_posts/2021-05-06-pixel-tests.md

```diff
@@ -0,0 +1,191 @@
+---
+title: Testing all the pixels
+author: mvo
+date: 2021-05-06 10:00
+tags: cockpit
+slug: pixel-testing
+category: tutorial
+comments: 'true'
+---
+
+The Cockpit integration tests can now contain "pixel tests". Such a
+test will take a screenshot with the browser and compare it with a
+reference.  The idea is that we can catch visual regressions much
+easier this way than if we would hunt for them in a purely manual
+fashion.
+
+## Preparing a repository for pixel tests
+
+A pixel test will take a screenshot of part of the Cockpit UI and
+compare it with a reference.  Thus, these reference images are
+important and play the biggest role.
+
+A large part of dealing with pixel tests will consequently consist of
+maintaining the reference images.  At the same time, we don't want to
+clog up our main source repository with them.  While the number and
+size of the reference images at any one point in time should not pose
+a problem, we will over time accumulate a history of them that we are
+afraid would dominate the source repository.
+
+Thus, the reference images are not stored in the source repository.
+Instead, we store them in an external repository that is linked into
+the source repository as a submodule.  That external repository
+doesn't keep any history and can be aggressively pruned.
+
+Developers are mostly isolated from this via the new
+`test/common/pixel-tests` tool.  But if you are familiar with git
+submodules, there should be no suprises for you here.
+
+A source repository needs to be prepared before it can store reference
+images in a external storage repository. You can let the tool do it by
+running
+
+```
+./test/common/pixel-tests init
+```
+
+and then committing the changes it has done to the source repository.
+(Those changes will be a new or modified .gitmodules file, and a new
+gitlink at test/reference.)
+
+## Adding a pixel test
+
+To add a pixel test to a test program, call the new "assert_pixels"
+function of the Browser class:
+
+     ...
```

Comment:
I suggest to drop the ... (they don't really provide much), or at least add a `def testSomeDialog(self):` header as a context.

You can use a ```py wrapper instead of indentation to get this printed more nicely.

Reply:
I have added the `def`.